### PR TITLE
(BIDS-2859) frontend - subscription to notifications modal

### PR DESCRIPTION
**Pull request by thib-wien**
### 1. Two dialog boxes to select the notifications to subscribe to

They can be [tried on the playground](https://bids-2859-frontend-subscript.beaconchain.pages.dev/playground).

image

### 2. `<BcNetworkSelector>` to allow users to choose a network (or just visualize a list)

Usage:
```HTML
<BcNetworkSelector v-model="selection" />
```
\- If `selection` is a `ref<ChainIDs>` then the user can select only one network.
\- If `selection` is a `ref<ChainIDs[]>` then the user can select several networks.
\- Or if you omit the v-model and instead give an array with `:readonly-networks="[...]"` in the props, then the component shows the networks in the array and the user cannot click.

image

### 3. `useRefBridge.ts` to manage the complexity of inputting and outputting on a v-model

`useRefBridge.ts` creates a two-way bridge between refs when they are of 2 different types. Example:
```TS
newRefOfTypeB = bridgeArrayRefs<TypeA,TypeB>(refOfTypeA)
```
There are 3 functions in the _.ts_ file to bridge 2 primitive variables, to bridge 2 arrays or to bridge 2 objects (two objects or one object and one array).
The values circulate back and forth between A and B, so the reactivity is preserved on both ends, the types are converted when they pass through, and infinite loops of A echoing B and so on are prevented.

If needed, the circulation on the bridge can be controlled by two methods embedded in the ref that the `bridgeRefs()` functions create, namely `.pauseBridgeFromNowOn()` and `.wakeupBridgeAtNextTick()`.

### 4. `isNetworkDisabled()` to know which networks must be hidden

This new utility function tells us which networks must be hidden in production mode and tells us to show them all in development mode. The function is returned by composable `useNetworkStore()`. Prototype:
```TS
isNetworkDisabled() : boolean
```
The purpose of this function is simple:
\- if the explorer is in its development mode, it always returns `false`,
\- if the explorer is in production mode, it returns `false` only for the network set in _.env_ and `true` for everything else.
This function will always return `false` in the future, when we are ready.

### 5. In development mode, more networks show up

In development mode, `availableNetworks` (returned by `useNetworkStore()`) contains now 4 networks. This allow us to make sure that what we do now is future-proof.
The consequences are for example:

Screenshot from 2024-07-06 12-54-24

### 6. `<BcInputNumber>`
This component is used to get numbers from the user, with a validation logic that allows us to output continuously on the v-model the number that the user types (`NaN` when the validation test fails) and auto-correct it on blur or Enter.
image

### 7. Updates and slighlty improves `<NotificationsManagementDashboards>`
- Makes it communicate for real with the API,
- Manages the saving of the settings from the subscription dialog boxes. From now on, the dialogs call `props.value.saveUserSettings(myNewPreferences)` each time they want to save the user preferences. The dialog boxes receive these props (@marcel-bitfly) :
```TS
{
  dashboardType: DashboardType,
  initialSettings: NotificationSettingsValidatorDashboard & NotificationSettingsAccountDashboard,
  saveUserSettings: (settings: NotificationSettingsValidatorDashboard & NotificationSettingsAccountDashboard) => void
}
```
- When hovering the subscriptions, shows a tooltip with the whole list:
image
- Displays human-readable captions (instead of object keys previously):
image
- Replaces ``<BcIconNetwork v-for="chainId of ..." />`` with `<BcNetworkSelector :readonly-networks="..." />`, so now the icons have a tooltip (useful for example to distinguish between Gnosis and Chiado because the icons are the same).
Screenshot from 2024-07-23 15-23-27

### CODE REVIEW

File: frontend/composables/useRefBridge.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ *     empty. You do not want the empty one to erase the initial data when the binding starts.
+ * @param origRef Ref that you want to bridge with the new ref that this function will create for you.
+ * @param origToCreated (optional) Callback/Arrow function that converts the value in the original ref into the type of the value in the created ref. If not provided, a basic conversion is performed, which is safe only between strings and numbers.
+ * @param createdToOrig (optional) Callback/Arrow function that converts the value in the created ref into the type of the value in the original ref. If not provided, a basic conversion is performed, which is safe only between strings and numbers.
+ * */
+export function usePrimitiveRefBridge<Torig, Tcreated> (origRef: Ref<Torig>|ModelRef<Torig>, origToCreated?: ConverterCallback<Torig, Tcreated>, createdToOrig?: ConverterCallback<Tcreated, Torig>) : Ref<Tcreated> {
+  return useObjectRefBridge<Torig, Tcreated>(origRef, origToCreated ?? convertPrimitiveByDefault, createdToOrig ?? convertPrimitiveByDefault)
+}
+
+function convertPrimitiveByDefault<TO> (from: any) : TO {
```
Comment (MauserBitfly) [inline frontend/composables/useRefBridge.ts:105]: would call `stringNumberConversion` that it is clear what it does, it would not be clear to me that the default mode would be to convert strings to number and vice versa.

File: frontend/composables/useRefBridge.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ *     the update triggers the first one again and so on.
+ *  3. At the moment the binding is created, it is typical that one array has initial data and the other is
+ *     empty. You do not want the empty one to erase the initial data when the binding starts.
+ * @param origRef Ref that you want to bridge with the new ref that this function will create for you.
+ * @param origToCreated (optional) Callback/Arrow function that converts an element in the original array into the type of the elements in the created array. If not provided, a basic conversion is performed, which is safe only between strings and numbers.
+ * @param createdToOrig (optional) Callback/Arrow function that converts an element in the created array into the type of the elements in the original array. If not provided, a basic conversion is performed, which is safe only between strings and numbers.
+ * */
+export function useArrayRefBridge<Torig, Tcreated> (origRef: Ref<Torig[]>|ModelRef<Torig[]>, origToCreated?: ConverterCallback<Torig, Tcreated>, createdToOrig?: ConverterCallback<Tcreated, Torig>) : Ref<Tcreated[]> {
+  const createdRef = ref<Tcreated[]>()
+  let pauseBack = false
```
Comment (MauserBitfly) [inline frontend/composables/useRefBridge.ts:21]: I think it must be possible to remove the pause variables (and the nextTick). (this is also valid for the ObjectRefBridge)

Example how it could work: (dummy code)
```ts
Origin changes ->
newOriginValue
const newCreatedValue = origToCreated(newOriginValue)
if(JSON.stringify(oldCreadedValue) === JSON.stringify(newCreatedValue)){
return
}
const testOriginValue = createdToOrign(newCreatedValue)
if(JSON.stringify(newOriginValue) === JSON.stringify(testOriginValue)){
warn('data conversion check failed', newOriginValue, newCreatedValue, testOriginValue)
return
}
createdRef.value = newCreatedValue
```

File: frontend/types/subscriptionModal.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  erc1155 = 'erc1155',
+
+  networks = 'networks',
+
+  ignoreSpam = 'ignore_spam',
+  ignore_spam = 'ignoreSpam'
+}
+
+/** for an option of type `number`: to mean that no value exists in the DB yet, set it to NaN; to mean that it is has a value but it is deactivated, set it to a negative value (for example 10% becomes -10) */
```
Comment (MauserBitfly) [inline frontend/types/subscriptionModal.ts:46]: Please add `TODO: replace Types once we have the type definitions from the API`

File: frontend/utils/translation.ts [hunk 4]
```diff
@@ -13,23 +16,37 @@ export function formatMultiPartSpan (t: ComposerTranslation, key: string, classe
...
 }
 
-// get all translations for an array
+/** returns an array of translations (with only one element if the translation data is a simple string) */
 export function tAll (t: ComposerTranslation, path : string, options?: any) : string[] {
   const list: string[] = []
-  let index = 0
-  for (let value = tOf(t, path, index); value; value = tOf(t, path, ++index, options)) {
-    list.push(value)
+  if (tD(t, path, options, NOT_FOUND) !== NOT_FOUND) {
```
Comment (MauserBitfly) [inline frontend/utils/translation.ts:41]:
```suggestion
  if (tD(t, path, option)) {
```
tD returns '' if not found so this check should be sufficient

File: frontend/utils/translation.ts [hunk 5]
```diff
@@ -13,23 +16,37 @@ export function formatMultiPartSpan (t: ComposerTranslation, key: string, classe
...
+    // the data is a string, we push the translation
+    list.push(tD(t, path, options))
+  } else {
+    // the data is an array
+    let index = 0
+    for (let value = tOfWithEOA(t, path, index, options); value !== END_OF_ARRAY; value = tOfWithEOA(t, path, ++index, options)) {
```
Comment (MauserBitfly) [inline frontend/utils/translation.ts:47]: I am pretty sure the previous version without the END_OF_ARRAY did also work

File: frontend/assets/css/main.scss [hunk 6]
```diff
@@ -27,6 +27,9 @@ a {
     pointer-events: none;
   }
 }
+hr {
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/assets/css/main.scss:30]: I did it because it had an effect. Without that CSS, `<hr>` was a standard light grey, not following our theme.

File: frontend/utils/translation.ts (hunk 4 above)
Comment (thib-wien) [reply frontend/utils/translation.ts:41]: The problem is that `''` caused `tAll()` to stop when a translation part was empty (which might happen for example because in some language this part is at a different place (like a verb in German that moves to the end)).
So I made this change to make sure that `tAll()` really stops at the end of the array and never before.

File: frontend/utils/translation.ts (hunk 5 above)
Comment (thib-wien) [reply frontend/utils/translation.ts:47]: The problem is that `''` caused `tAll()` to stop when a translation part was empty (which might happen for example because in some language this part is at a different place (like a verb in German that moves to the end)).
So I made this change to make sure that `tAll()` really stops at the end of the array and never before.

File: frontend/types/subscriptionModal.ts (hunk 3 above)
Comment (thib-wien) [reply frontend/types/subscriptionModal.ts:46]: Done

File: frontend/composables/useRefBridge.ts (hunk 2 above)
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts:21]: Unfortunately not. It is similar to my initial solution but it lead to bugs. Sometimes, what we send at one end of the bridge might be slightly changed by the conversion call-backs so the output looks different and then these `JSON.stringify(newOriginValue) === JSON.stringify(testOriginValue)` cause infinite loops.
This is too restrictive, it would make the bridge unusable for some use cases. We can make the bridge more versatile if we ignore the content of the refs.

Maybe you mean that you are worried that 
``TS
pauseBack = true
nextTick(() => { pauseBack = false })
```
might not always work.
The only source of bug here would be that the Back watcher fires two ticks later, thus escaping the pause and starting a loop.
I am not sure that this would match Vue's specifications. I think that Vue ensures that all watchers run within the same tick, so the template is rendered with consistent data at the end of the tick:
https://vuejs.org/guide/essentials/watchers#callback-flush-timing
> By default, *a watcher's callback is called* after parent component updates (if any), and *before the owner component's DOM updates*. This means if you attempt to access the owner component's own DOM inside a watcher callback, the DOM will be in a pre-update state.

File: frontend/composables/useRefBridge.ts (hunk 1 above)
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts:105]: Thanks!
Done

File: frontend/components/notifications/SubscriptionRow.vue [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  .caption {
+    margin-right: 3px;
+  }
+
+  .info {
+    margin-left: 3px;
+    margin-right: 3px;
+  }
+  .gem {
+    margin-left: 3px;
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionRow.vue:171]: Fixed

File: frontend/components/notifications/SubscriptionRow.vue [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+      <BcNetworkSelector v-model="networkSelectorState" :class="deactivationClass" />
+    </div>
+  </div>
+</template>
+
+<style scoped lang="scss">
+@use "~/assets/css/fonts.scss";
+
+.deactivated {
+  opacity: 0.6;
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionRow.vue:152]: Done.

File: frontend/composables/useRefBridge.ts (hunk 2 above)
Comment (MauserBitfly) [reply frontend/composables/useRefBridge.ts:21]: but if the return conversion returns a different result and we skip this change wouldn't that then make the data invalid

if origin != (origin -> created -> origin) then there is something wrong in the data conversion. Otherwise we would get a wrong origin if the created changes later

File: frontend/components/bc/NetworkSelector.vue [hunk 9]
```diff
@@ -0,0 +1,56 @@
+<script setup lang="ts">
+import type { MultiBarItem } from '~/types/multiBar'
+import { IconNetwork } from '#components'
+import { ChainInfo, ChainIDs } from '~/types/network'
+
+const props = defineProps<{
+  displayOnly?: boolean
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:7]: OK!
Done

File: frontend/utils/translation.ts (hunk 5 above)
Comment (MauserBitfly) [reply frontend/utils/translation.ts:47]: I don't see the need for empty strings in the translations. 
Example: 
English would have the bold text in the middle
["This is my ", "bold, "test"]
If in another language the bold text is at the end then it would only have two text parts
["Am Ende ist der Text", "fett"]

... I don't see a use case where we would have an empty string in the middle

File: frontend/components/bc/NetworkSelector.vue [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+}>()
+
+const { availableNetworks, isNetworkDisabled } = useNetworkStore()
+
+/** If the v-model is:
+ *  - A ChainIDs: only one network can be selected by the user. Prop `:display-only` must be false or omitted.
+ *  - An array of ChainIDs:
+ *    - and prop `:display-only` is `false`/omitted: several networks can be selected by the user,
+ *    - and prop `:display-only` is `true`: the networks in the array are shown to the user but they are unclickable. */
+const liveState = defineModel<ChainIDs|ChainIDs[]>({ required: false })
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:17]: > Would prefer to have the model to the outside be called 'selection' like for the other Bars's.

Done.

> The selection should only be used in editable mode. I would suggest to have an optional prop 'availableNetworks?: ChainIDs[]' ... if this is set then it is used for the button list, else the one from the useNetworkStore

This complicates the code a a lot as you can see.
Also, it was making sense that the list of networks in the readonly mode was coming from the v-model `selection` because we are displaying the selection of the user.
But I changed it as you asked. I suppose that the way to use the component is now more standard. I trust my teacher.

File: frontend/components/bc/toggle/MultiBar.vue [hunk 11]
```diff
@@ -3,45 +3,57 @@ import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
 import { type MultiBarItem } from '~/types/multiBar'
 
 interface Props {
-  icons: MultiBarItem[]
+  displayMode?: boolean,
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/toggle/MultiBar.vue:6]: Done

File: frontend/components/bc/NetworkSelector.vue [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+const selection = Array.isArray(liveState.value)
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:19]: Using only `useObjectRefBridge` would force the code to contain conversion callbacks.
Not clearer I think.

File: frontend/composables/useRefBridge.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  const stopperBack = watch(createdRef, () => {
+    if (pauseBack) { return }
+    const CasO = (createdRef.value !== undefined) ? createdToOrig(createdRef.value) : undefined as unknown as Torig
+    origRef.value = CasO
+    pauseForth = true
+    nextTick(() => { pauseForth = false })
+  }, { deep: true })
+
+  onUnmounted(() => {
+    stopperBack()
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts:80]: Vue clean watchers only when they are started in script setup.
I am not sure that the composable is considered as part of script setup.
If the bridge is created in a function, then for sure Vue will not stop the watchers.

File: frontend/components/bc/NetworkSelector.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:17]: Fixed I think.

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+const isOptionAvailable = (key: string) => props.value?.premiumUser || !(key in DefaultValueOfValidatorOptionsNeedingPremium || key in DefaultValueOfAccountOptionsNeedingPremium)
+
+function closeDialog () : void {
+  if (lastSaveFailed) {
+    // second chance: we try not to lose what the user has set
+    sendUserPreferencesToAPI()
+  }
+  dialogRef?.value.close(true)
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:142]: OK, fixed.

File: frontend/types/subscriptionModal.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
+import { ChainIDs } from './network'
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/types/subscriptionModal.ts:1]: Done

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+  .title {
+    @include fonts.dialog_header;
+    text-align: center;
+    margin-bottom: var(--padding-large);
+  }
+
+  .explanation {
+    margin-bottom: var(--padding);
+    @include fonts.small_text;
+    opacity: 0.6;
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:217]: OK, I set it to ` --text-color-discreet`.
I had used opacity because the discreet text looks almost black with the light color theme. I do not see the point then.

I contacted Philip to check that it is really what we want.

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+      />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.proposedBlock" :t-path="tPath+'proposed_block'" :lacks-premium-subscription="!isOptionAvailable('proposedBlock')" class="row" />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.upcomingProposal" :t-path="tPath+'upcoming_proposal'" :lacks-premium-subscription="!isOptionAvailable('upcomingProposal')" class="row" />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.syncCommittee" :t-path="tPath+'sync_committee'" :lacks-premium-subscription="!isOptionAvailable('syncCommittee')" class="row" />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.withdrawn" :t-path="tPath+'withdrawn'" :lacks-premium-subscription="!isOptionAvailable('withdrawn')" class="row" />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.slashed" :t-path="tPath+'slashed'" :lacks-premium-subscription="!isOptionAvailable('slashed')" class="row" />
+      <NotificationsSubscriptionRow v-model="modifiableOptions.realTime" :t-path="tPath+'real_time'" :lacks-premium-subscription="!isOptionAvailable('realTime')" class="row" />
+      <hr>
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:179]: `<hr>` has the color given by Figma I believe.
But [it seems that this tag is more or less deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hr) so I replaced it with a `<div/>`

File: frontend/components/notifications/SubscriptionRow.vue [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+    <BcTooltip v-if="tooltipLines[0]" :fit-content="true">
+      <FontAwesomeIcon :icon="faInfoCircle" class="info" />
+      <template #tooltip>
+        <BcMiniParser :input="tooltipLines" class="tt-content" />
+      </template>
+    </BcTooltip>
+    <BcPremiumGem v-if="lacksPremiumSubscription" class="gem" />
+    <div v-if="type != 'networks'" class="right">
+      <div v-if="type == 'amount' || type == 'percent'" class="input">
+        <InputText
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionRow.vue:123]: This would not remove the validation logic because it is used by the bridge to make sure that the real-time changes that we send to the API are valid.

For example, when the user types `952` in the `%` input, our specification makes us output something to the API immediately (I mean with debouncing) but the automatic correction of `<InputNumber>` will kick in only if the user unfocuses.
The callbacks of the bridge need the correcting functions.

File: frontend/components/notifications/SubscriptionRow.vue [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+interface CheckboxAndText {
+  check: boolean,
+  text: string
+}
+
+const state = defineModel<CheckboxAndNumber|ChainIDs[]>({ required: true })
+let networkSelectorState: ModelRef<ChainIDs[]>
+let checkboxAndText: Ref<CheckboxAndText> | undefined
+
+// bridging the v-model (CheckboxAndNumber | ChainIDs[]) with the data of this component (CheckboxAndText | ChainIDs[])
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionRow.vue:29]: I forgot the the policy of Bitfly is "make your code as difficult to understand as possible".
I removed my comments separating the script into clear purposes.

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+const { fetch, setTimeout } = useCustomFetch()
+const toast = useBcToast()
+const { networkInfo } = useNetworkStore()
+
+const tPath = ref('')
+let originalSettings = {} as AllPossibleOptions
+const modifiableOptions = ref({} as ModifiableOptions)
+const allCheckbox = ref({ check: false } as CheckboxAndNumber)
+
+const debouncer = useDebounceValue<number>(0, MinimumTimeBetweenAPIcalls)
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:40]: What would be the goal/benefit?

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+        category: props.value!.accountSub ? 'accounts' : 'validators',
+        subscriptions: output
+      }
+    })
+  } catch {
+    response = undefined
+  }
+  if (!response || response.error) {
+    toast.showError({ summary: t('notifications.subscriptions.error_title'), group: t('notifications.subscriptions.error_group'), detail: t('notifications.subscriptions.error_message') })
+    lastSaveFailed = true // we will try again when the dialog closes
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:129]: OK, I removed the last-chance save when the dialog gets closed.

File: frontend/components/bc/NetworkSelector.vue [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+  ? useArrayRefBridge<ChainIDs, string>(liveState as Ref<ChainIDs[]>)
+  : usePrimitiveRefBridge<ChainIDs, string>(liveState as Ref<ChainIDs>)
+
+const buttons = shallowRef<MultiBarItem[]>([])
+
+if (props.displayOnly) {
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:25]: Indeed, a side effect of you other idea is to fix this little lack.

File: frontend/components/bc/NetworkSelector.vue [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+  watch(liveState as Ref<ChainIDs[]>, updateButtons, { immediate: true })
+} else {
+  watch(availableNetworks, updateButtons, { immediate: true })
+}
+
+function updateButtons (source: ChainIDs[]) : void {
+  buttons.value = []
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:32]: Done

File: frontend/components/bc/MiniParser.vue [hunk 24]
```diff
@@ -0,0 +1,226 @@
+<script setup lang="ts">
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/MiniParser.vue:1]: I think I understand your opposition if this is something that you never saw in the companies you worked for: it looks non-standard.
Does it make it bad?

Maybe we could sit and talk about it rationally? I am not against good practice, I just want to understand what disadvantage is stronger than the benefits.

>  that much complexity

Your example demonstrates that a parser is less complex. You suggest:
```json
"title": "My title",
"list": {
  "0": ["first list item",  "a bold text",  "another normal text"],
  "1": ["second list item",  "a bold text",  "another normal text"],
}
```
but the json with a parser is simply:
```json
[
  "My title",
  "- first list item *a bold text* another normal text",
  "- second list item *a bold text* another normal text"
]
```

> will be harder to maintain / translate / adapt in the future

The current situation consists in hard-coded spans structures in the templates and list structures in the translation files, and some of them might be invalid when we translate to other languages. (+ we need hacks to silence eslint warnings)

You suggest to create a component for lists. It would come in addition to `formatMultiSpan` for bold and italic. Then we will we create a component for links...
Basically those things are "hard-coded parsers" and each of them requires complicated structures in the translation files and take the risk to be invalid with some languages and inject HTML into the VDOM.

I would like to understand rationally what is the problem with a parser that brings simplicity and cleanliness to the code and the json.

File: frontend/types/network.ts [hunk 25]
```diff
@@ -31,7 +31,7 @@ export enum ChainIDs {
 }
 
 export interface ChainInfoFields {
-  name: string,
+  name: string[],
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/types/network.ts:34]: OK, I reverted all the `.name.join(' ')` to `.name`.
Now, the buttons of _DashboardCreationNetworkMask.vue_ use a new member:
```TS
const buttonList = computed(() => {
  list.push({
    text: ChainInfo[chainId].nameParts[0],
    subText: ChainInfo[chainId].nameParts[1],
    ...
```

File: frontend/composables/useRefBridge.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts:80]: OK, removed.

File: frontend/types/subscriptionModal.ts [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+
+/** translates the names of the members of our objects to/from the names used by the API  // TODO: write here the identifiers actually used by the API */
+export enum SubscriptionJSONfields {
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/types/subscriptionModal.ts]: > This mapping makes the whole process quit complex ... I think it would be more straight forward to use the backend API structs directly.

I agree (I did not know how to do and I did not like my solutions).
Changed.
Now the component use the JSON of _types/api/notifications.ts_ (as it is not available to us yet, I copied its prototypes into _types/subscriptionModal.ts_)

File: frontend/composables/useRefBridge.ts (hunk 2 above)
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts]: What you propose probably makes sense in a context that you have in mind but that I fail to imagine.

In which context would it be good that we check that the pair of conversion functions form a bijective mapping between the two refs?

To my mind, the contract that we have with the developer using the bridge is this one:
- each time ref `A` changes, sends its data to ref `B` through a conversion function,
- each time ref `B` changes, sends its data to ref `A` through a conversion function.

This is what we promise and do.

Forcing the pair of functions to form a bijective mapping would limit the use cases. I am not sure that it would bring a benefit to add a verification of the bijection to the contract. What I am sure about is that it would annoy colleagues using our bridge in some cases that they need to tackle.

I understand that forcing the bijection would prevent bugs for distracted developers, but it would do so at the expense of use cases for developers knowing what they are doing.

File: frontend/utils/translation.ts (hunk 4 above)
Comment (thib-wien) [reply frontend/utils/translation.ts:41]: You are right, we would return an empty array, but I do not find consistent to return `[]` (meaning "no translation found") when there is a translation (which an empty string but still, the field in the JSON is not missing so I find it more consistent to return `[""]`).

File: frontend/utils/translation.ts (hunk 5 above)
Comment (thib-wien) [reply frontend/utils/translation.ts:47]: > I don't see a use case where we would have an empty string in the middle

Consider this two sentences with bold and links:
- As we like both, the [chocolate](chocolink) that you bring _**may**_ be dark or with milk.
- Da wir beide mögen, _**darf**_ die [Schokolade](chocolink), die du bringst, dunkel oder mit Milch sein.

Their arrays need empty elements:

```TS
["As we like both, the", "chocolate", "that you bring", "may", "be dark or with milk."]
```
```TS
["Da wir beide mögen", "", "", "darf", "die", "Schokolade", "die du bringst, dunkel oder mit Milch sein"]
```
And even if we don't talk about chocolate on beaconcha.in, I am convinced that stopping the scan of the array at the first empty element was a bug.

**Comment by thib-wien** [comment]
@MauserBitfly `NotificationsManagementDashboards.vue` is now getting its data from the real endpoint that Lucca created. As the API response is different from what you imagined, the creation of computed object `wrappedDashboardGroups` that you fill with the response is just a bit more complex. You can see [here](https://github.com/gobitfly/beaconchain/blob/794f8ffac0fa9b8a8256f98552daa8dd02713d8c/frontend/components/notifications/management/NotificationsManagementDashboards.vue#L75) how it is filled now (some members are now calculated instead of just copied from the response).

There is one information that I don't properly fill yet: the name of the dashboard.

File: frontend/components/notifications/SubscriptionDialog.vue (hunk 20 above)
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/notifications/SubscriptionDialog.vue]: to update fast if the user only updates one item but prevents spamming if the user clicks checkboxes multiple times / fast in sequence. 

the second parameter of the bounce function makes this possible if set to true

File: frontend/components/notifications/management/NotificationsManagementDashboards.vue [hunk 27]
```diff
@@ -56,6 +118,26 @@ const onEdit = (col: 'delete' | 'subscriptions' | 'webhook' | 'networks', row: N
   }
 }
 
+async function saveUserSettings (settingsAndContext: SettingsWithContext) {
+  let response: ApiErrorResponse | undefined
+  try {
+    setTimeout(TimeoutForSavingFailures)
```
Comment (MauserBitfly) [inline frontend/components/notifications/management/NotificationsManagementDashboards.vue:124]: there is no resolve function for this setTimeout ... can we remove it?

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 28]
```diff
@@ -0,0 +1,264 @@
+<script setup lang="ts">
```
Comment (MauserBitfly) [inline frontend/components/notifications/SubscriptionDialog.vue:1]: could you please move the two subscription components in the management folder so that all managing stuff is together? 
(sorry I did not see this before)

File: frontend/components/notifications/SubscriptionDialog.vue (hunk 17 above)
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/notifications/SubscriptionDialog.vue]: better, but couldn't you also set the border top of the 'top' row instead of adding an extra div? (if too complicated please ignore)

File: frontend/composables/useRefBridge.ts (hunk 2 above)
Comment (MauserBitfly) [reply frontend/composables/useRefBridge.ts]: Hmm I still think the bridge should be save in both ways. And not create different outcomes when converted back and forth. 
Here is my example:
-> 1. We load data 
-> 2.  data is converted into input data
-> 3. user changes data
-> 4. data is converted to backend data
...
-> 5. data is again loaded from the backend
-> 6. data is converted to input data  ..... and this has to be the same as 3

But we can leave it for now and adapt if needed.

File: frontend/components/playground/MiniParser.vue [hunk 29]
```diff
@@ -0,0 +1,52 @@
+<script setup lang="ts">
```
Comment (MauserBitfly) [inline frontend/components/playground/MiniParser.vue:1]: Maybe also move the MiniParser.vue in your other branch ;-)

**Review by MauserBitfly — COMMENTED** [review]
Thanks for all the refactoring ... nicely done. 

Added some more comment.

File: frontend/components/notifications/SubscriptionDialog.vue (hunk 20 above)
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue]: 🫡 OK, easy.
Done.

File: frontend/components/notifications/SubscriptionDialog.vue (hunk 17 above)
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue]: As you can see here
image
the bottom of the rows is not in the middle between two rows.

It might be possible to change the whole CSS to do what you want but should I spend time trying to change something that works and is simple?

File: frontend/composables/useRefBridge.ts (hunk 2 above)
Comment (thib-wien) [reply frontend/composables/useRefBridge.ts]: Yes, we can add a boolean parameter in the future that triggers a bijectivity check for distracted programmers who need it,
and keep it optional so uses cases like the example I cited (210° == -150°).

File: frontend/components/bc/NetworkSelector.vue [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+ *  Otherwise, give a v-model. If the v-model is
+ *   a ChainIDs: only one network can be selected by the user,
+ *   an array of ChainIDs: several networks can be selected by the user */
+const selection = defineModel<ChainIDs|ChainIDs[]>({ required: false })
+
+let barSelection: Ref<string> | Ref<string[]>
+if (props.readonlyNetworks) {
+  barSelection = ref<string[]>([])
+} else
+  if (Array.isArray(selection.value)) {
```
[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/bc/NetworkSelector.vue:23]: 🫡

File: frontend/components/notifications/management/NotificationsManagementDashboards.vue (hunk 27 above)
Comment (thib-wien) [reply frontend/components/notifications/management/NotificationsManagementDashboards.vue:124]: It is actually a function of _useCutomFetch()_ that is imported [here](https://github.com/gobitfly/beaconchain/blob/c1e459f4398bc57dafcee09cea9b9c201025f89b/frontend/components/notifications/management/NotificationsManagementDashboards.vue#L44). We discussed it [here](https://github.com/gobitfly/beaconchain/pull/528#discussion_r1668652319).

Should I really remove it or just rename it to avoid confusion?

File: frontend/components/notifications/SubscriptionDialog.vue (hunk 28 above)
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue:1]: OK 🫡

File: frontend/components/notifications/SubscriptionDialog.vue [hunk 31]
```diff
@@ -0,0 +1,232 @@
...
+      switch (typeof originalSettings[key]) {
+        case 'boolean' : output[apiKey] = entry[1].check; break
+        case 'number' : output[apiKey] = entry[1].num * (entry[1].check ? 1 : -1); break
+      }
+    }
+  }
+  // now we send the data
+  let response: ApiErrorResponse | undefined
+  try {
+    setTimeout(TimeoutForSavingFailures)
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/notifications/SubscriptionDialog.vue]: Sorry missed this discussion in my second review. 

Why do you need the new `setTimeout` function ... you could just pass the timeout as a signal parameter with the options if you want to override the default…

[the remark this answers is not in the document]
Comment (thib-wien) [reply frontend/components/notifications/SubscriptionDialog.vue]: Indeed.
Fixed.

**Review by MauserBitfly — APPROVED** [review]
LGTM
